File: 203 VB Inherited Events (Chien)/class Meute.vb
Imports System.Collections.ObjectModel


Class Meute(Of T As Animal)
    Private ReadOnly mCol As Collection(Of T)

    Sub New()
        mCol = New Collection(Of T)
    End Sub

    Sub New(ByVal a1 As T)
        mCol = New Collection(Of T)
        mCol.Add(a1)
    End Sub

    Sub Add(ByVal a1 As T)
        mCol.Add(a1)
    End Sub

    Sub Enerver()
        For Each a As T In mCol
            a.Enerver()
        Next
    End Sub

End Class


'' Custom Event Handler in VB
'Class Events
'    Private Event PreDrawEvent As EventHandler

'    Private Custom Event OnDraw As EventHandler
'        AddHandler(ByVal value As EventHandler)
'            SyncLock PreDrawEvent
'                AddHandler PreDrawEvent, value
'            End SyncLock
'        End AddHandler
'        RemoveHandler(ByVal value As EventHandler)
'            SyncLock PreDrawEvent
'                RemoveHandler PreDrawEvent, value
'            End SyncLock
'        End RemoveHandler
'    End Event
'
'End Class

File: Net6/Net6Converted 100-209/203 CS Inherited Events (Chien)/class Meute.cs
// 2021-09-19   PV  VS2022; Net6

using System;
using System.Collections.ObjectModel;

#pragma warning disable IDE0051 // Remove unused private members

namespace CS203
{
    internal class Meute<T> where T : Animal
    {
        private readonly Collection<T> mCol;

        public Meute()
        {
            mCol = new Collection<T>();
        }

        public Meute(T a1)
        {
            // Using a Collection Initializer
            mCol = new Collection<T> { a1 };
        }

        public void Add(T a1)
        {
            mCol.Add(a1);
        }

        public void Enerver()
        {
            foreach (T a in mCol)
            {
                a.Enerver();
            }
        }
    }

    // Custom Event Handler in C#
    internal class Events
    {
        private event EventHandler PreDrawEvent;

        private event EventHandler OnDraw
        {
            add
            {
                lock (PreDrawEvent)
                {
                    PreDrawEvent += value;
                }
            }
            remove
            {
                lock (PreDrawEvent)
                {
                    PreDrawEvent -= value;
                }
            }
        }
    }
}